rules/policy.go

```diff
@@ -78,23 +77,20 @@ func filterNets(mixedCIDRs []string, ipVersion uint8) (filtered []string, filter
 	return
 }
 
-func (r *DefaultRuleRenderer) ProtoRuleToIptablesRules(pRule *proto.Rule, ipVersion uint8) []iptables.Rule {
-	// Filter the CIDRs to the IP version that we're rendering.  In general, we should have an
-	// explicit IP version in the rule and all CIDRs should match it (and calicoctl, for
-	// example, enforces that).  However, we try to handle a rule gracefully if it's missing a
-	// version.
-	//
-	// We do that by rendering the rule, filtered to only have CIDRs of the right version,
-	// unless filtering the rule would completely remove one of its match fields.
-	//
-	// That handles the mainline case well, where the IP version is missing but the rule is
-	// otherwise consistent since we'll render the rule only for the matching version.
-	//
-	// It also handles rules like "allow from 10.0.0.1,feed::beef" in an intuitive way.  Only
-	// rules of the form "allow from 10.0.0.1,feed::beef to 10.0.0.2" will get filtered out,
-	// and only for IPv6, where there's no obvious meaning to the rule.
-	ruleCopy := *pRule
+func FilterRuleToIPVersion(ipVersion uint8, pRule *proto.Rule) *proto.Rule {
+	ruleCopy := pRule
```

Comment:
Since `pRule` is a pointer to a rule, this isn't making a copy of the rule, it's making a copy of the pointer.  That means that the changes below will modify the input rule, which will cause problems.

Couple of ways round that:

- Change this to `ruleCopy = *pRule` and change the return to be `return &ruleCopy` (this is 100% safe in go, unlike in C)
- Change the argument to be `pRule proto.Rule` and the result to be `proto.Rule`; then the copy semantics will be explicit in the signature.